expression/builtin_compare.go

```diff
@@ -2429,7 +2431,18 @@ func CompareString(sctx sessionctx.Context, lhsArg, rhsArg Expression, lhsRow, r
 	if isNull0 || isNull1 {
 		return compareNull(isNull0, isNull1), true, nil
 	}
-	return int64(types.CompareString(arg0, arg1)), false, nil
+	ret := compareStringWithTrim(arg0, lhsArg.GetType(), arg1, rhsArg.GetType())
+	return int64(ret), false, nil
+}
+
+func compareStringWithTrim(lhs string, lhsFt *types.FieldType, rhs string, rhsFt *types.FieldType) int {
+	if lhsFt.Collate != charset.CollationBin {
```

Comment:
We need some comments here according to to MySQL manual to describe why we need to trim here.